packages/hub/test/session.test.ts

```diff
@@ -0,0 +1,117 @@
+      ['sets a duration', { duration: 12000 }, { duration: 12000 }],
+      ['does not use custom duration for browser env', { duration: 12000, isBrowser: true }, { duration: undefined }],
+      [
+        'does not set a negative duration',
+        { timestamp: 10, started: 100 },
+        // TODO(abhi): What should the behaviour here be?
+        { duration: 0, timestamp: expect.any(String), started: expect.any(String) },
+      ],
+      [
+        'sets duration based on timestamp and started',
+        { timestamp: 100, started: 10 },
+        { duration: 90, timestamp: expect.any(String), started: expect.any(String) },
+      ],
+      [
+        'sets a release',
+        { release: 'f1557994979ecd969963f53c27ca946379d721f3' },
+        { attrs: { release: 'f1557994979ecd969963f53c27ca946379d721f3' } },
+      ],
+      ['sets an environment', { environment: 'staging' }, { attrs: { environment: 'staging' } }],
+      ['sets an ipAddress', { ipAddress: '0.0.0.0' }, { attrs: { ip_address: '0.0.0.0' } }],
+      [
+        'overwrites user ip_address did with custom ipAddress',
+        { ipAddress: '0.0.0.0', user: { ip_address: '1.1.1.1' } },
+        { attrs: { ip_address: '0.0.0.0' } },
+      ],
+      ['sets an userAgent', { userAgent: 'Mozilla/5.0' }, { attrs: { user_agent: 'Mozilla/5.0' } }],
+      ['sets errors', { errors: 3 }, { errors: 3 }],
+      ['sets status', { status: SessionStatus.Crashed }, { status: SessionStatus.Crashed }],
+    ];
+
+    test.each(table)('%s', (...test) => {
+      // duration and timestamp can vary after session update, so let's expect anything unless
+      // the out variable in a test explicitly refers to it.
+      const DEFAULT_OUT = { duration: expect.any(Number), timestamp: expect.any(String) };
+
+      const session = new Session();
+      const initSessionProps = session.toJSON();
+
+      session.update(test[1]);
+      expect(session.toJSON()).toEqual({ ...initSessionProps, ...DEFAULT_OUT, ...test[2] });
+    });
+  });
+
+  describe('close', () => {
+    it('exits a normal session', () => {
+      const session = new Session();
+      const mockUpdate = jest.spyOn(session, 'update');
+      expect(mockUpdate).toHaveBeenCalledTimes(0);
+      session.close();
+      expect(mockUpdate).toHaveBeenCalledTimes(1);
+      expect(mockUpdate).toHaveBeenLastCalledWith({ status: SessionStatus.Exited });
+    });
+
+    it('updates session status when give status', () => {
+      const session = new Session();
+      const mockUpdate = jest.spyOn(session, 'update');
+
+      session.close(SessionStatus.Crashed);
+      expect(mockUpdate).toHaveBeenCalledTimes(1);
+      expect(mockUpdate).toHaveBeenLastCalledWith({ status: SessionStatus.Crashed });
+    });
+
+    it('only changes status ok', () => {
+      const session = new Session();
+      session.status = SessionStatus.Abnormal;
+      const mockUpdate = jest.spyOn(session, 'update');
+
```

Comment:
```suggestion
expect(mockUpdate).toHaveBeenCalledTimes(0);
```

Again just noticing the lack of symmetry among the tests.

Not sure this assertion adds much value though, it feels like testing the test framework as nothing could possibly have called the function at this point.